src/device/random.jl

```diff
@@ -0,0 +1,181 @@
+    blockId = blockIdx().x + (blockIdx().y - 1) * gridDim().x +
+                             (blockIdx().z - 1) * gridDim().x * gridDim().y
+    z = xorshift(z | blockId%UInt32)
+
+    return z
+end
+
+# NOTE: @propagate_inbounds for when we know the passed seed contains at least 32 elements
+# XXX: without @inline @propagate_inbounds, we get local memory accesses (e.g. with `rand!`)
+
+#const TausShift1 = CuConstantMemory{UInt32}((6, 2, 13, 3))
+#const TausShift2 = CuConstantMemory{UInt32}((13, 27, 21, 12))
+#const TausShift3 = CuConstantMemory{UInt32}((18, 2, 7, 13))
+#const TausOffset = CuConstantMemory{UInt32}((4294967294, 4294967288, 4294967280, 4294967168))
+# XXX: constant memory isn't supported yet, so let's resort to llvmcall
+for (name, vals) in ("TausShift1" => (6, 2, 13, 3),
+                     "TausShift2" => (13, 27, 21, 12),
+                     "TausShift3" => (18, 2, 7, 13),
+                     "TausOffset" => (4294967294, 4294967288, 4294967280, 4294967168))
+    @eval @inline function $(Symbol(name))()
+        ptr = Base.llvmcall(
+            $("""@$(name) = weak addrspace($(AS.Constant)) global [4 x i32] [i32 $(vals[1]), i32 $(vals[2]), i32 $(vals[3]), i32 $(vals[4])], align 4
+                define i8 addrspace($(AS.Constant))* @entry() #0 {
+                    %ptr = getelementptr inbounds [4 x i32], [4 x i32] addrspace($(AS.Constant))* @$(name), i64 0, i64 0
+                    %untyped_ptr = bitcast i32 addrspace($(AS.Constant))* %ptr to i8 addrspace($(AS.Constant))*
+                    ret i8 addrspace($(AS.Constant))* %untyped_ptr
+                }
+                attributes #0 = { alwaysinline }
+            """, "entry"), LLVMPtr{UInt32, AS.Constant}, Tuple{})
+        CuDeviceArray((4,), ptr)
+    end
+end
+
+function TausStep(z::Unsigned, S1::Integer, S2::Integer, S3::Integer, M::Unsigned)
+    b = (((z << S1) ⊻ z) >> S2)
+    return (((z & M) << S3) ⊻ b)
+end
+
+"""
+    Random.rand(rng::SharedTauswortheGenerator, T)
+
+Generate a byte of random data using the on-device Tausworthe generator. This function
+performs synchronization when updating the shared RNG state, so it should be called
+uniformly in a thread block (i.e. do not call it from a branch that may diverge within a
+block).
+"""
+function Random.rand(rng::SharedTauswortheGenerator, ::Type{UInt32})
+    # 0-indexed so that we can bitwise and instead of mod1
+    tid0 = threadIdx().x - 1 + (threadIdx().y - 1) * blockDim().x +
+                               (threadIdx().z - 1) * blockDim().x * blockDim().y
+
+    i = tid0&31 + 1
+    j = tid0&3  + 1
+    @inbounds begin
+        z = rng.state[i]
+        sync_threads()  # XXX: this implies that rand() cannot be called from a branch
+        if z == 0
+            z = initial_state(rng.seed)
+        end
+        S1, S2, S3, M = TausShift1()[j], TausShift2()[j], TausShift3()[j], TausOffset()[j]
+        rng.state[i] = TausStep(z, S1, S2, S3, M)
```

Comment:
Do we not need a `sync_threads()` here?

Reply:
Yes, good catch. Too bad we need synchronization here, because that more or less means we can't use this function from branches.